Add Table of Contents block

SeanDS opened the pull request:
_Please see the stale PR #15426 and issue #11047 for background. This PR is simply a rebase of the changes on #15426 onto the current master - all credit to @ashwin-pc and [sorta brilliant](https://sortabrilliant.com/guidepost/). As of writing this contains only the changes made there, with only tiny changes to comments. There are likely some bugs probably still need ironed out. The text below is copied from that PR._

Code contributions in this commit entirely made by ashwin-pc, originally based on the
"Guidepost" block by sorta brilliant (https://sortabrilliant.com/guidepost/).

## Description

Adding a Table of Contents block. #11047

This block is based off a similar block plugin [Guidepost](https://sortabrilliant.com/guidepost/) but rebuilt to essentially behave similar to Table of Contents as seen on other major platforms such as Google Docs and Word.

The block auto populates links based on the heading hierarchy of the post with the option to either:

auto-sync changes
![autosync](https://user-images.githubusercontent.com/20453492/57105234-419ce000-6cf8-11e9-91ef-54fa11c10b23.gif)

or to sync changes when manually triggered.
![refresh](https://user-images.githubusercontent.com/20453492/57105251-49f51b00-6cf8-11e9-870e-71ca0dbcd8ac.gif)

When auto sync is turned off the block also allows you to manually edit the link names similar to other Table of Contents implementations.
![edit](https://user-images.githubusercontent.com/20453492/57105299-6d1fca80-6cf8-11e9-8e2b-410244d69881.gif)

## How has this been tested?
Passed all unit tests using `npm run test` and all end to end tests using `npm run test-e2e`. new Unit Tests will have to be written for the new block though.

## Screenshots
Already given above.

## Types of changes
New feature (non-breaking change which adds functionality).

## Checklist:
   * [x]  My code is tested.
   * [x]  My code follows the WordPress code style.
   * [x]  My code follows the accessibility standards.
   * [x]  My code has proper inline documentation.
   * [x]  I've included developer documentation if appropriate.

SeanDS committed:
Copy changes from pull request #15426 (https://github.com/WordPress/gutenberg/pull/15426).

Adds Table of Contents block to the editor.

Code contributions in this commit entirely made by ashwin-pc, originally based on the
"Guidepost" block by sorta brilliant (https://sortabrilliant.com/guidepost/).

SeanDS commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+		let anchor = contentDiv.textContent || contentDiv.innerText || '';
+
+		if ( anchor !== '' && anchor.indexOf( '#' ) === -1 ) {
+			anchor = '#' + cleanForSlug( anchor );
```
I noticed a bug: this function doesn't strip off HTML tags so when the title contains HTML these end up in the anchor:

![Screenshot_20200320_110901](https://user-images.githubusercontent.com/5225190/77154196-257e6a80-6a93-11ea-9e2c-540d65ca9521.png)

Any suggestions for a built-in Gutenberg/JavaScript function to use to properly slugify the anchor?

SeanDS replied:
I've implemented your code, but wouldn't `getTextContent` from `RichText` do the same job?

I wonder if we want to allow the HTML style from the header to copy over into the table of contents entry too.

ZebulanStanphill commented:
With regard to the auto-sync behavior, I think it's a bit of a weird way to handle custom-named links. I think it would be better if the auto-sync was always on, and you could simply convert the block to a List block if you wanted to start customizing the names. This would be a one-way conversion, of course, but re-enabling the auto-sync behavior on the Table of Contents block would wipe out your customizations anyway, so it doesn't matter.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+	// Add anchors to any headings that don't have one.
+	getHeadingBlocks().forEach( function( heading, key ) {
+		const headingAnchorEmpty =
+			typeof heading.attributes.anchor === 'undefined' ||
```
I think this could be simplified:
```suggestion
			heading.attributes.anchor === undefined ||
```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+			typeof heading.attributes.anchor === 'undefined' ||
+			heading.attributes.anchor === '';
+		const headingContentEmpty =
+			typeof heading.attributes.content === 'undefined' ||
```
```suggestion
			heading.attributes.content === undefined ||
```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+	const returnValue = [];
+
+	array.forEach( function( heading, key ) {
+		if ( typeof heading.content === 'undefined' ) {
```
```suggestion
		if ( heading.content === undefined ) {
```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+			// If it does and the next level is greater than the current level,
+			// the next iteration becomes a child of the current interation.
+			if (
+				typeof array[ key + 1 ] !== 'undefined' &&
```
```suggestion
				array[ key + 1 ] !== undefined &&
```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+		contentDiv.innerHTML = anchorContent;
+		let anchor = contentDiv.textContent || contentDiv.innerText || '';
+
+		if ( anchor !== '' && anchor.indexOf( '#' ) === -1 ) {
```
This can be made more clear using [`String.prototype.includes`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes):
```suggestion
		if ( anchor !== '' && ! anchor.includes( '#' ) ) {
```

SeanDS committed:
Update packages/block-library/src/table-of-contents/utils.js

Co-Authored-By: Zebulan Stanphill <zebulanstanphill@protonmail.com>

SeanDS committed:
Update packages/block-library/src/table-of-contents/utils.js

Co-Authored-By: Zebulan Stanphill <zebulanstanphill@protonmail.com>

SeanDS committed:
Update packages/block-library/src/table-of-contents/utils.js

Co-Authored-By: Zebulan Stanphill <zebulanstanphill@protonmail.com>

SeanDS committed:
Update packages/block-library/src/table-of-contents/utils.js

Co-Authored-By: Zebulan Stanphill <zebulanstanphill@protonmail.com>

SeanDS committed:
Update packages/block-library/src/table-of-contents/utils.js

Co-Authored-By: Zebulan Stanphill <zebulanstanphill@protonmail.com>

SeanDS commented:
@ZebulanStanphill 

> With regard to the auto-sync behavior, I think it's a bit of a weird way to handle custom-named links. I think it would be better if the auto-sync was always on, and you could simply convert the block to a List block if you wanted to start customizing the names. This would be a one-way conversion, of course, but re-enabling the auto-sync behavior on the Table of Contents block would wipe out your customizations anyway, so it doesn't matter.

Yes I agree, I didn't like that change from the original Guidepost plugin behaviour, which just didn't let you edit the contents. I like your idea of allowing it to be changed one way to a list though.

paaljoachim commented:
I just tested this out by using gutenberg.run (it took a few tries to get it to run)
I really liked what I saw.

Going through the process.

Finding the block.
![Screen Shot 2020-03-21 at 08 04 09](https://user-images.githubusercontent.com/5323259/77221550-d98efc80-6b4a-11ea-8c7f-62a1b3ea2f99.png)

Block settings
![Screen Shot 2020-03-21 at 08 04 23](https://user-images.githubusercontent.com/5323259/77221553-e7dd1880-6b4a-11ea-98ac-fc951ff0f931.png)

Turning off Auto Sync
![Screen Shot 2020-03-21 at 08 04 48](https://user-images.githubusercontent.com/5323259/77221560-f3c8da80-6b4a-11ea-87a8-24b36e99ae48.png)

Clicking Auto Sync
![Screen Shot 2020-03-21 at 08 05 43](https://user-images.githubusercontent.com/5323259/77221576-122ed600-6b4b-11ea-9e17-c12eb4de9a65.png)

Auto syncing removes any renaming to the Header links.

It would be nice to auto sync and at the same time keep the name customizations in place without these having to revert to the original heading name.

Perhaps a check box. Something like this.
![Content-Block-Not-Syncing-Custom-Names](https://user-images.githubusercontent.com/5323259/77221688-2b845200-6b4c-11ea-9108-7a57f1c9fc1c.jpg)

Anyhow we should have someone from design go through this PR.
@mapk @jasmussen

SeanDS commented:
@paaljoachim thanks for checking it out, though almost all credit should go to the previous PR author and the original authors of Guidepost!

I'd indeed be keen to hear from design reviewers before making any major changes.

mapk commented:
I tested this today and noticed a bug. I wasn't sure how I got it, but I dropped some ipsum on my page, created a few headings in the middle, then came back to the top to create another heading. I ended up with this:

![Screen Shot 2020-03-23 at 4 29 38 PM](https://user-images.githubusercontent.com/617986/77372756-a6ec2c00-6d23-11ea-88c5-ed7c6c57ec12.png)

As you can see, the first heading listed in my Table of Contents block isn't on my page. I'm not sure why it's there or how to get rid of it. I have "auto sync" turned "on."

paaljoachim commented:
I retested with using gutenberg.run

Went here: https://lipsum.com/
Added 5 paragraphs.
Added the Table of Contents Block.

Backend result:
![Screen Shot 2020-03-24 at 12 14 41](https://user-images.githubusercontent.com/5323259/77419639-0cdebf00-6dc9-11ea-94dd-d199bffdf112.png)

Frontend result:
![Screen Shot 2020-03-24 at 12 15 14](https://user-images.githubusercontent.com/5323259/77419704-208a2580-6dc9-11ea-9e26-476009193a01.png)

draganescu commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,134 @@
+		this.refresh = this.refresh.bind( this );
+	}
+
+	toggleAttribute( propName ) {
```
This method is used in a single place. Is it useful instead of simply setting the attribute to ! attribute?

draganescu commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,134 @@
+				<ListLevel
+					edit={ true }
+					attributes={ attributes }
+					setAttributes={ setAttributes }
```
Maybe instead of passing down `setAttributes` we could use an `onChange` handler and have that defined in the controlling component instead which in turn would call `setAttributes` ?

draganescu commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,134 @@
+				}
+				<ListLevel
+					edit={ true }
+					attributes={ attributes }
```
in `ListLevel` we only use `headings` and `autosync` why not pass them explicitly?

draganescu reviewed (changes requested):
Great work rebasing and refreshing @SeanDS, generally this looks and works very nice. Left some code related comments. Also:

- when turning off AutoSync and then editing the text of the links the post fails to update

![toc-no-auto-upade](https://user-images.githubusercontent.com/107534/77428377-a796c800-6de0-11ea-811d-d8e0910cca66.gif)


- the `LinkControl` when clicking on the link button in the block's toolbar is empty. This is tricky as the `LinkControl` does not support anchors yet.

<img width="508" alt="Screenshot 2020-03-24 at 14 59 34" src="https://user-images.githubusercontent.com/107534/77428410-b2e9f380-6de0-11ea-8106-d2625e8f3250.png">

- we should remove some of the formatting options available (code, image, strikethrough)

<img width="504" alt="Screenshot 2020-03-24 at 15 02 31" src="https://user-images.githubusercontent.com/107534/77428425-ba110180-6de0-11ea-9304-c6e9951beebd.png">

ZebulanStanphill commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+	} );
+
+	// Return boolean value from length.
+	return !! +changedHeadings.length;
```
I also don't see why the `+` is necessary.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+export function linearToNestedList( array, index = 0 ) {
+	const returnValue = [];
+
+	array.forEach( function( heading, key ) {
```
I think a `.reduce` might actually make more sense in this case.

draganescu replied:
Yes, even better, as we're building that `returnValue` thing :)

SeanDS replied:
But the loop needs to look ahead to the next value (with the `key` parameter) to check if it will return a value. I didn't write this (it's uplifted from the old PR or the original plugin) but the use of `forEach` here seems appropriate to me.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+
+export function convertHeadingBlocksToAttributes( headingBlocks ) {
+	return headingBlocks.map( function( heading ) {
+		const level = heading.attributes.level.toString();
```
Why is `level` a string instead of a number?

talldan replied:
This would probably be because it's sourced from an HTML attribute. Block attributes that are sourced from HTML attributes have to be strings.

ZebulanStanphill replied:
@talldan [But it's not sourced from an HTML attribute, and it's stored as a number.](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/heading/block.json#L15)

talldan replied:
@ZebulanStanphill The value comes from the heading block as a `number`, but is converted to a `string` here so that it can be stored/sourced as an attribute in the table of contents block using a data attribute:
https://github.com/WordPress/gutenberg/pull/21040/files#diff-ca24950398fffc6e682a847d2d23f966R31

draganescu replied:
Aha! nice, but at that position in code it's hard to figure. Maybe convert later before assignment ...

talldan replied:
Yep, a comment explaining the conversion might also be a solution.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/editor.scss:
```diff
@@ -0,0 +1,21 @@
+	}
+
+	p {
+		opacity: 0.5;
```
What is the purpose of this style?

SeanDS replied:
Not used. Removed.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/ListLevel.js:
```diff
@@ -0,0 +1,74 @@
+
+export default function ListLevel( props ) {
+	const { edit, attributes, setAttributes } = props;
+	let childnodes = null;
```
This variable should probably be called `childNodes`. (There are also some variables called `childnode` that should probably be `childNode`.)

SeanDS committed:
Merge branch 'master' into add/table-of-contents

SeanDS committed:
Merge branch 'master' into add/table-of-contents

SeanDS committed:
Fix variable name capitalisation

SeanDS committed:
Rename function

SeanDS committed:
Reword description

SeanDS committed:
Change argument from array to headingsList

SeanDS committed:
Change variable name

SeanDS committed:
Add comment

SeanDS committed:
Get rid of autosync (users should now convert to list if they want to edit the contents)

SeanDS committed:
Add ability to transform into list; remove unused ListLevel props

SeanDS committed:
Update table-of-contents block test configuration

SeanDS committed:
Simplify expression

SeanDS committed:
Remove unused function

SeanDS committed:
Remove unused style

SeanDS commented:
Thanks @ZebulanStanphill and @draganescu. I removed the auto-sync switch because I prefer the idea to allow converting one-way to a list. I implemented the transform for converting to a list so this is now possible. Removing the auto-sync behaviour also fixed a lot of the suggested changes. Please have another look at the code and let me know if there are more suggestions.

I am still unsure about the `cleanForSlug` bug I found above. Does anyone know of a good way to strip off special characters from the URL anchors?

talldan reviewed (commented):
This turned into a pretty massive code review! I hope it's helpful.

I think there's a lot to be careful about with this block, both in how it synchronizes it's data and how it updates other blocks.

Would be good to get some design feedback on how to handle headings that don't have anchors. I'll add the appropriate label.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/editor.scss:
```diff
@@ -0,0 +1,17 @@
+			&:focus {
+				box-shadow: none;
+			}
+		}
```
Why are these styles needed?

SeanDS replied:
Not really useful - removed.

SeanDS committed:
Remove unnecessary style

SeanDS committed:
Rename TOCEdit to TableOfContentsEdit

SeanDS committed:
Apply suggestions from code review

Co-Authored-By: Daniel Richards <daniel.p.richards@gmail.com>

SeanDS committed:
Merge branch 'add/table-of-contents' of github.com:SeanDS/gutenberg into add/table-of-contents

SeanDS committed:
Remove non-existent import

SeanDS committed:
Make imports explicit

SeanDS committed:
Remove unused function

SeanDS committed:
Change unsubscribe function to class property

SeanDS committed:
Change JSON.stringify comparison to Lodash's isEqual

SeanDS committed:
Turns out refresh() is required

SeanDS committed:
Remove unnecessary state setting

SeanDS committed:
Don't change state on save

SeanDS committed:
Change behaviour to only add links if there are anchors specified by the user

SeanDS committed:
Newline

SeanDS committed:
Replace anchor with explicit key in map since anchor can now sometimes be empty

SeanDS committed:
Update test data

SeanDS commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,134 @@
+		this.state.wpDataUnsubscribe();
+	}
+
+	componentDidUpdate( prevProps, prevState ) {
```
This is no longer updating state so (I think?) doesn't trigger a re-render. It might still be possible to merge into `subscribe` though - what do you think?

SeanDS commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,134 @@
+
+	refresh() {
+		const { setAttributes } = this.props;
+		const headings = Utils.getPageHeadings();
```
I got rid of `refresh` altogether - it was left over from some removed code. Now `getPageHeadings` (now called `getHeadingsList`) is only used in `componentDidMount`.

SeanDS commented on packages/block-library/src/table-of-contents/utils.js:
```diff
@@ -0,0 +1,135 @@
+	return returnValue;
+}
+
+export function getPageHeadings() {
```
Since this (now called `getHeadingsList`) is used twice in `componentDidMount`, it seems useful to keep it.

SeanDS commented on packages/block-library/src/table-of-contents/save.js:
```diff
@@ -0,0 +1,27 @@
+		return null;
+	}
+
+	Utils.updateHeadingBlockAnchors();
```
I really like this suggestion, since it removes any guess work and requires explicit user input for entries in the table of contents to become links. I've changed the code to work this way, and it has simplified it a great deal. Now the block does not change any other components.

SeanDS commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,134 @@
+	componentDidMount() {
+		const { attributes, setAttributes } = this.props;
+		const headings = attributes.headings || [];
+		const wpDataUnsubscribe = subscribe( () => {
```
I've dealt with your second point by removing any ability to change other blocks altogether. The first point is probably still a problem. Please take a look at the latest version and let me know if you have a suggestion on avoiding re-renders on every trigger.

SeanDS commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,134 @@
+			this.setState( { pageHeadings } );
+		} );
+
+		setAttributes( { headings } );
```
I think I had to keep that there to make the block work when newly created, but I haven't checked again since changing a bunch of other things. I can experiment removing it to see if it still behaves.

SeanDS commented:
Thanks for everyone's continuing enthusiasm!

I've implemented most of the suggestions from @talldan and @ZebulanStanphill and left comments for some suggestions I wasn't clear on or to explain where I've taken a different approach. In this update I've removed the code that updates the headers in the document entirely, so the block now only listens for changes to headers. As a consequence, it no longer generates anchors on headers. Headers without anchors are shown in the table of contents without a link.

Please take another look and let me know what you think. I am sure there are still some changes required, particularly the new code I've added, and I'm not convinced the component update functions (`componentDidMount`, `componentDidUpdate`) are optimally triggering renders and state updates. Any feedback here would be great.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/block.json:
```diff
@@ -0,0 +1,17 @@
+			}
+		}
+	}
+}
```
This should have a trailing newline, as per the [root `.editorconfig` file](https://github.com/WordPress/gutenberg/blob/master/.editorconfig).
```suggestion
}

```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/index.js:
```diff
@@ -0,0 +1,28 @@
+export const settings = {
+	title: __( 'Table of Contents' ),
+	description: __(
+		'Add a list of internal links allowing your readers to quickly navigate around.'
```
I think this description needs to be updated since the block no longer automatically adds anchors to the headings.

SeanDS replied:
True. Maybe "Add a list of headers allowing your readers to quickly navigate through your post." instead, and somewhere else make it obvious to users how to make them links (maybe with popovers or similar)?

ZebulanStanphill replied:
I'd say "headings" rather than "headers". Also, the contributor guide recommends [avoiding describing features as "allowing" something](https://developer.wordpress.org/block-editor/contributors/document/copy-guide/#two-cut-phrases-that-inflate-your-word-count-without-actually-adding-meaning).

How's this?

"A list of headings to summarize your post. Add HTML anchors to Heading blocks to automatically link them here."

ZebulanStanphill commented on packages/block-library/src/table-of-contents/ListLevel.js:
```diff
@@ -0,0 +1,43 @@
+ * WordPress dependencies
+ */
+
+export default function ListLevel( props ) {
```
I think `ListItem` would be a more appropriate name.

SeanDS committed:
Update packages/block-library/src/table-of-contents/block.json

Co-Authored-By: Zebulan Stanphill <zebulanstanphill@protonmail.com>

SeanDS committed:
Rename ListLevel to ListItem for clarity

ZebulanStanphill commented on packages/block-library/src/table-of-contents/ListItem.js:
```diff
@@ -0,0 +1,43 @@
+	let childNodes = null;
+
+	if ( children ) {
+		childNodes = children.map( function( childNode, index ) {
```
`childNodes` isn't used except inside the `if` statement, and it's never reassigned.
```suggestion

	if ( children ) {
		const childNodes = children.map( function( childNode, index ) {
```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/ListItem.js:
```diff
@@ -0,0 +1,43 @@
+ */
+
+export default function ListItem( props ) {
+	const { children, noWrapList = false } = props;
```
```suggestion
export default function ListItem( { children, noWrapList = false } ) {
```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/ListItem.js:
```diff
@@ -0,0 +1,43 @@
+/**
+ * WordPress dependencies
+ */
+
```
This comment should be removed.
```suggestion
```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,72 @@
+/**
+ * External dependencies
+ */
+
```
This newline should be removed.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,72 @@
+ * Internal dependencies
+ */
+import { getHeadingsList, linearToNestedHeadingList } from './utils';
+import ListItem from './ListItem';
```
These should come after the WordPress dependencies, not before.

ZebulanStanphill commented on packages/block-library/src/table-of-contents/edit.js:
```diff
@@ -0,0 +1,72 @@
+			return (
+				<p>
+					{ __(
+						'Start adding heading blocks to see a Table of Contents here'
```
```suggestion
						'Start adding Heading blocks to create a table of contents here.'
```

ZebulanStanphill commented on packages/block-library/src/table-of-contents/ListItem.js:
```diff
@@ -0,0 +1,43 @@
+
+			const entry = anchor ? (
+				<a
+					className="blocks-table-of-contents-entry"
```
This class name should follow the BEM-ish style used elsewhere in Gutenberg:
```suggestion
					className="wp-block-table-of-contents__entry"
```

ZebulanStanphill replied:
(The tests will have to be updated.)

ZebulanStanphill commented on packages/block-library/src/table-of-contents/ListItem.js:
```diff
@@ -0,0 +1,43 @@
+				</a>
+			) : (
+				<span
+					className="blocks-table-of-contents-entry"
```
```suggestion
					className="wp-block-table-of-contents__entry"
```

ZebulanStanphill commented:
I made a derivative of this PR that refactors the edit component to use hooks, as well as apply various bits of polish I wanted: #21234. Let me know what you think! :smile:

SeanDS commented:
So should development continue here or in #21234?

ZebulanStanphill commented:
@SeanDS I figure that if the hooks approach ends up not being viable for whatever reason, we can go back to this PR. So feel free to continue polishing this PR as a backup. But otherwise, I think we should continue development in #21234.

draganescu committed:
Update packages/block-library/src/table-of-contents/ListItem.js

Co-Authored-By: Zebulan Stanphill <zebulanstanphill@protonmail.com>

SeanDS commented:
Given the awesome work @ZebulanStanphill has done on #21234 so far, I'm closing this in favour of the many improvements made there.
